USM: enable Go TLS monitoring by default

Yumasi opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

This PR enables USM's Go TLS monitoring by default.

### Motivation

### Describe how to test/QA your changes

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Yumasi committed:
USM: enable Go TLS monitoring by default

guyarb commented on pkg/config/setup/system_probe.go:
```diff
@@ -241,7 +241,7 @@ func InitSystemProbeConfig(cfg pkgconfigmodel.Config) {
 
 	// For backward compatibility
-	cfg.BindEnv(join(smNS, "enable_go_tls_support"))
+	cfg.BindEnvAndSetDefault(join(smNS, "enable_go_tls_support"), true)
 	cfg.BindEnv(join(smNS, "tls", "go", "enabled"))
```
Since we have backward compatibility, the change should be done in `adjust_usm`
Only if the flags weren't set, then we should set to true

guyarb commented on pkg/network/config/config_test.go:
```diff
@@ -1463,34 +1463,34 @@ func TestUSMTLSNativeEnabled(t *testing.T) {
 		cfg := New()
 
-		require.True(t, cfg.EnableGoTLSSupport)
+		require.False(t, cfg.EnableGoTLSSupport)
 	})
 
 	t.Run("via deprecated ENV variable", func(t *testing.T) {
 		mock.NewSystemProbe(t)
-		t.Setenv("DD_SERVICE_MONITORING_CONFIG_ENABLE_GO_TLS_SUPPORT", "true")
+		t.Setenv("DD_SERVICE_MONITORING_CONFIG_ENABLE_GO_TLS_SUPPORT", "false")
 		cfg := New()
 
-		require.True(t, cfg.EnableGoTLSSupport)
+		require.False(t, cfg.EnableGoTLSSupport)
 	})
 
 	t.Run("via YAML", func(t *testing.T) {
 		mockSystemProbe := mock.NewSystemProbe(t)
-		mockSystemProbe.SetWithoutSource("service_monitoring_config.tls.go.enabled", true)
+		mockSystemProbe.SetWithoutSource("service_monitoring_config.tls.go.enabled", false)
 		cfg := New()
 
-		require.True(t, cfg.EnableGoTLSSupport)
+		require.False(t, cfg.EnableGoTLSSupport)
 	})
 
 	t.Run("via ENV variable", func(t *testing.T) {
 		mock.NewSystemProbe(t)
-		t.Setenv("DD_SERVICE_MONITORING_CONFIG_TLS_GO_ENABLED", "true")
+		t.Setenv("DD_SERVICE_MONITORING_CONFIG_TLS_GO_ENABLED", "false")
 		cfg := New()
 
-		require.True(t, cfg.EnableGoTLSSupport)
+		require.False(t, cfg.EnableGoTLSSupport)
```
Instead of changing the tests, I'd add a scenario for `true` and `false`

Yumasi replied:
I am not sure I get it. I changed the tests because, to me, their value is in proving that setting the config via yaml or env works. If I leave them unchanged, the tests no longer prove it works because we can't differentiate if the config value is true because of the yaml/env or because of the default value.

Yumasi replied:
So, I noticed now that making the change of the default value in `adjust_usm`, alongside the ones for HTTP, HTTP2, etc..., does not require actually changing the tests at all. Although, it means having a test for a "default value" that checks it is false (which is true only if USM is not enabled), like we have for HTTP config.

Yumasi replied:
So, it turns out that with the change made in `adjust_usm.go`, the tests are no longer affected (they still have a false default value, like we have for other protocols like the HTTP config tests). So I reverted the changes to them.

Yumasi committed:
fix CR

Yumasi committed:
revert BindEnvAndSetDefault

Yumasi committed:
revert tests changes

The change in behaviour no longer affect the tests with the default
value being set in adjust_usm.go

guyarb commented on cmd/system-probe/config/adjust_usm.go:
```diff
@@ -20,6 +20,7 @@ func adjustUSM(cfg model.Config) {
 	if cfg.GetBool(smNS("enabled")) {
 		applyDefault(cfg, netNS("enable_http_monitoring"), true)
 		applyDefault(cfg, netNS("enable_https_monitoring"), true)
+		applyDefault(cfg, spNS("enable_go_tls_support"), true)
```
did you test the change?
a.k.a running USM without configuring go-tls and checking it was enabled?

Yumasi replied:
Good catch. I admit I forgot to do so, I should have caught this.

Yumasi committed:
fix namespace is adjustUSM

Yumasi committed:
add tests to check default value depending on whether USM is enabled or not

guyarb commented on cmd/system-probe/config/adjust_usm.go:
```diff
@@ -20,7 +20,7 @@ func adjustUSM(cfg model.Config) {
 		applyDefault(cfg, netNS("enable_http_monitoring"), true)
 		applyDefault(cfg, netNS("enable_https_monitoring"), true)
-		applyDefault(cfg, spNS("enable_go_tls_support"), true)
+		applyDefault(cfg, smNS("enable_go_tls_support"), true)
```
why using the deprecated value?

Yumasi committed:
use the current config option instead of the deprecated one in adjustUSM

guyarb commented on cmd/system-probe/config/adjust_usm.go:
```diff
@@ -20,6 +20,7 @@ func adjustUSM(cfg model.Config) {
 	if cfg.GetBool(smNS("enabled")) {
 		applyDefault(cfg, netNS("enable_http_monitoring"), true)
 		applyDefault(cfg, netNS("enable_https_monitoring"), true)
+		applyDefault(cfg, smNS("tls", "go", "enabled"), true)
```
that's not the place to put it.
you're breaking the backward compatibility

Yumasi replied:
Could you explain? I don't get it.

guyarb replied:
Try to run usm with the follow configuration:

```yaml
service_monitoring_config:
  enabled: true
  enable_go_tls_support: false
```

Yumasi replied:
Changed it to only apply the default if the deprecated option is not set.

Yumasi committed:
preserve backward compatibility of GoTLS options

Yumasi committed:
simplify code

Yumasi committed:
update tests

Yumasi committed:
update releasenote

Yumasi committed:
fix tracer tests

Yumasi committed:
fix inventoryagent tests

GustavoCaso reviewed (approved):
LGTM for the file owned by @DataDog/agent-shared-components

Yumasi committed:
fix kafka monitor test

Yumasi committed:
fix shared library tests

Yumasi committed:
fix http2 test

Yumasi committed:
fix nodejs test

Yumasi committed:
fix kafka test

Yumasi committed:
Merge branch 'main' into guillaume.pagnoux/enable-gotls-by-default

Yumasi committed:
Merge branch 'main' into guillaume.pagnoux/enable-gotls-by-default

Yumasi committed:
update releasenotes

Yumasi commented:
/merge
